packages/core/src/lib/filter/subscription.ts

```diff
@@ -0,0 +1,286 @@
+
+      log.info("Ping successful");
+    } catch (error) {
+      log.error("Error pinging: ", error);
+      throw new Error("Error pinging: " + error);
+    }
+  }
+
+  async unsubscribeAll(): Promise<void> {
+    const stream = await this.newStream(this.peer);
+
+    const request = FilterSubscribeRpc.createUnsubscribeAllRequest(
+      this.pubsubTopic
+    );
+
+    try {
+      const res = await pipe(
+        [request.encode()],
+        lp.encode,
+        stream,
+        lp.decode,
+        async (source) => await all(source)
+      );
+
+      if (!res || !res.length) {
+        throw Error(
+          `No response received for request ${request.requestId}: ${res}`
+        );
+      }
+
+      const { statusCode, requestId, statusDesc } =
+        FilterSubscribeResponse.decode(res[0].slice());
+
+      if (statusCode < 200 || statusCode >= 300) {
+        throw new Error(
+          `Filter unsubscribe all request ${requestId} failed with status code ${statusCode}: ${statusDesc}`
+        );
+      }
+
+      for (const decoder of this.decoders.values()) {
+        this.removeEventListener(decoder.contentTopic);
+      }
+      this.decoders.clear();
+      log.info("Unsubscribed from all content topics");
+    } catch (error) {
+      throw new Error("Error unsubscribing from all content topics: " + error);
+    }
+  }
+
+  addEventListener<K extends string | number>(
+    contentTopic: K,
+    listener: EventHandler<SubscriptionEventMap[K]> | null,
+    options?: boolean | AddEventListenerOptions | undefined
+  ): void {
+    if (typeof contentTopic !== "string")
+      throw new Error(
+        `Invalid event type. Expected content topic which should be a string but received ${contentTopic} of the type ${typeof contentTopic}`
+      );
+
+    const decoder = this.decoders.get(contentTopic);
+    if (!decoder)
+      throw new Error("No decoder found for content topic " + contentTopic);
+    if (decoder.pubsubTopic !== this.pubsubTopic)
+      throw new Error(
+        `Invalid decoder received. Expected decoder for pubsub topic ${this.pubsubTopic} but received decoder for pubsub topic ${decoder.pubsubTopic}`
+      );
+
+    super.addEventListener(contentTopic, listener, options);
+  }
+
+  async processMessage(message: WakuMessage): Promise<void> {
+    const { contentTopic } = message;
+    if (!contentTopic) {
+      log.info("Message has no content topic, skipping");
+      return;
+    }
+
+    const decoder = this.decoders.get(contentTopic);
+    if (decoder) {
+      try {
+        const decodedMessage = await decoder.fromProtoObj(
+          this.pubsubTopic,
+          message as IProtoMessage
+        );
+        if (!decodedMessage) {
+          throw new Error("Decoding failed");
+        }
+        this.dispatchEvent(
+          new CustomEvent<IDecodedMessage>(decoder.contentTopic, {
+            detail: decodedMessage
+          })
+        );
+      } catch (e) {
+        log.error("Error decoding message", e);
```

Comment:
This function and `Filter.onRequest` should be updated to allow listeners being notified about errors as I mentioned here https://github.com/waku-org/js-waku/pull/1682/files#r1387338041

Reply:
Let's split it into a follow-up task